packages/components/src/Chip/Chip.tsx

```diff
@@ -31,29 +33,73 @@ export const Chip = ({
   const prefix = useChildComponent(children, d => d.type === Chip.Prefix);
   const suffix = useChildComponent(children, d => d.type === Chip.Suffix);
 
+  const [labelRef, labelFullyVisible] = useInView<HTMLSpanElement>();
+  const [headingRef, headingFullyVisible] = useInView<HTMLSpanElement>();
+  const tooltipMessage = getTooltipMessage(
+    labelFullyVisible,
+    headingFullyVisible,
+    label,
+    heading,
+  );
+
   return (
-    <button
-      className={classes}
-      onClick={(ev: React.MouseEvent<HTMLButtonElement>) =>
-        onClick && onClick(value, ev)
-      }
-      tabIndex={tabIndex}
-      onKeyDown={onKeyDown}
-      aria-label={ariaLabel || label}
-      disabled={disabled}
-      role={role}
-      type="button"
-    >
-      {prefix}
-      <Typography size="base" fontWeight="medium">
-        {heading}
-      </Typography>
-      {heading && <span className={styles.chipBar} />}
-      <Typography size="base">{label}</Typography>
-      {suffix}
-    </button>
+    <Tooltip message={tooltipMessage}>
```

Comment:
I know this is throwing a wrench in things pretty late, and I know this is a pretty big abstract ask (so feel free to turn it down), but is there a way to pull apart these two pieces and combine them based on a prop? Like to separate the concern of Chip, and then a Chip with a Tooltip because it's too long?

Brainstorming here for ways to accomplish that, but could some of this functionality be pulled into a local hook, which is then enabled via prop so the functionality is close by, but not wrapping everything?

I'm just thinking down the road where maybe we want to repurpose Chip again for something else, and you've already established the super awesome pattern that allows Chip to be conditionally repurposed for greater things. Or, do we hold on to that idea until that situation actually comes up, and then do the work then. That's equally valid.

Reply:
I hear ya. unfortunately that's how I initially had it with a 

```
function ChipWrapper(props: {children: ReactElement, showTooltip: boolean}):JSX.Element {
if (props.showTooltip) {
  return <Tooltip {...whateverPropsGoHere}>{props.children}</Tooltip>
}

return props.children
}
```

where we invoked it with

```
<ChipWrapper>
  <Chip {...props}/>
</ChipWrapper>
```

the problem was that the tooltip kept getting removed and re-added infinitely, causing a really bad time overall. the fix was to always have the tooltip elements present, but not show them when the message is empty which is why that tooltip change was added in this PR

Reply:
ideally, yes I would love to have that setup where Chip doesn't always have a Tooltip wrapping it, because I do agree it's a bit goofy to include it when you know full well it has no use. the interactions between `useInView` and `Tooltip` were just not having it :(